In [1]:
#import libraries
import pandas as pd
import numpy as np

from sklearn.metrics import pairwise_distances

In [2]:
#import dataset
book_df = pd.read_csv("book.csv",encoding='ISO-8859-1',index_col=0)
book_df.head()

User.ID                                         Book.Title  Book.Rating
1   276726                                Classical Mythology            5
2   276729                                       Clara Callan            3
3   276729                               Decision in Normandy            6
4   276736  Flu: The Story of the Great Influenza Pandemic...            8
5   276737                             The Mummies of Urumchi            6

## EDA

In [3]:
#Duplicate redords
book_df[book_df.iloc[:,0:2].duplicated()]

User.ID                   Book.Title  Book.Rating
2002   278418     White Oleander : A Novel            5
3793     1211                  Black House            9
4746     2033               The First Time            2
5052     2152    Le nouveau soleil de Teur            7
5220     2276  In the Beauty of the Lilies           10
7440     3757          The Magician's Tale            7
9946   162052                         Rose            8

In [4]:
#Delete the duplicate records
book_df.drop(book_df[book_df.iloc[:,0:2].duplicated()].index, inplace=True)
book_df.reset_index(drop=True,inplace=True)

In [5]:
#total number of records
book_df.shape

(9993, 3)

In [6]:
#Total number of unique users
len(book_df["User.ID"].unique())

2182

In [7]:
#Total number of unique books
len(book_df["Book.Title"].unique())

9659

In [8]:
#Create pivot table using userID and booktitle
user_book_df = book_df.pivot(index="User.ID",
                                 columns='Book.Title',
                                 values='Book.Rating')

In [9]:
#Fill na values with 0
user_book_df.fillna(0,inplace=True)

In [10]:
user_book_df

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
User.ID                                                                        
8                            0.0                                         0.0   
9                            0.0                                         0.0   
10                           0.0                                         0.0   
12                           0.0                                         0.0   
14                           0.0                                         0.0   
...                          ...                                         ...   
278846                       0.0                                         0.0   
278849                       0.0                                         0.0   
278851                       0.0                                         0.0   
278852                       0.0                                         0.0   
278854                       0.0                                         0.0   

Book.Title   Repairing PC Drives &amp  '48  \
User.ID                                      
8                                 0.0  0.0   
9                                 0.0  0.0   
10                                0.0  0.0   
12                                0.0  0.0   
14                                0.0  0.0   
...                               ...  ...   
278846                            0.0  0.0   
278849                            0.0  0.0   
278851                            0.0  0.0   
278852                            0.0  0.0   
278854                            0.0  0.0   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
User.ID                                                                             
8                                                         0.0                       
9                                                         0.0                       
10                                                        0.0                       
12                                                        0.0                       
14                                                        0.0                       
...                                                       ...                       
278846                                                    0.0                       
278849                                                    0.0                       
278851                                                    0.0                       
278852                                                    0.0                       
278854                                                    0.0                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
User.ID                                                                    
8                                                         0.0              
9                                                         0.0              
10                                                        0.0              
12                                                        0.0              
14                                                        0.0              
...                                                       ...              
278846                                                    0.0              
278849                                                    0.0              
278851                                                    0.0              
278852                                                    0.0              
278854                                                    0.0              

Book.Title  01-01-00: A Novel of the Millennium  \
User.ID                                           
8                                           0.0   
9                                           0.0   
10                                          0.0   
12        

## Similarity Calculation

In [11]:
#cosine similarity
user_sim = 1- pairwise_distances(user_book_df.values, metric="cosine")
user_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [12]:
#Dataframe with cosine scores
np.fill_diagonal(user_sim,0)
user_sim_df = pd.DataFrame(user_sim)
user_sim_df.index = user_book_df.index
user_sim_df.columns = user_book_df.index
user_sim_df

User.ID  8       9       10      12      14      16      17      19      \
User.ID                                                                   
8           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
9           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
10          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
12          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
14          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
278846      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278849      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278851      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278852      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278854      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

User.ID  22      26      ...  278831  278832  278836  278843  278844  278846  \
User.ID                  ...                                                   
8           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
9           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
10          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
12          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
14          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
278846      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278849      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278851      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278852      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278854      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

User.ID  278849  278851  278852  278854  
User.ID                                  
8           0.0     0.0     0.0     0.0  
9           0.0     0.0     0.0     0.0  
10          0.0     0.0     0.0     0.0  
12          0.0     0.0     0.0     0.0  
14          0.0     0.0     0.0     0.0  
...         ...     ...     ...     ...  
278846      0.0     0.0     0.0     0.0  
278849      0.0     0.0     0.0     0.0  
278851      0.0     0.0     0.0     0.0  
278852      0.0     0.0     0.0     0.0  
278854      0.0     0.0     0.0     0.0  

[2182 rows x 2182 columns]

In [13]:
#Now similar users can found from the above datframe, example
user_sim_df.idxmax(axis=1)[0:15]

User.ID
8          8
9          8
10         8
12         8
14         8
16         8
17         8
19    278418
22         8
26         8
32         8
39         8
42         8
44         8
51      3757
dtype: int64

## From the above results we can say users 19 and 278418 have similarities, and also users 51 and 3757